### PR TITLE
[WIP] Tracking the unsolved variables that was assigned `!` type

### PR DESCRIPTION
**Pull request by blitzerr**
Addresses #66173

### CODE REVIEW

File: compiler/rustc_session/src/lint/builtin.rs [hunk 1]
```diff
@@ -2555,6 +2555,13 @@ declare_lint! {
     @feature_gate = sym::unsafe_block_in_unsafe_fn;
 }
 
+declare_lint! {
+    //TODO: Add explanation.
+    pub FALL_BACK_TO_NEVER_TYPE,
+    Deny,
+    "Unresolved variable might fall back to never_type `!`"
```
Comment (bjorn3) [inline compiler/rustc_session/src/lint/builtin.rs:2562]:
```suggestion
    "Unresolved variable might fall back to never_type `!` in the future"
```

Should this be a future compatibility lint?

File: src/test/ui/never_type/never_type_lint.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#![feature(never_type_fallback)]
+
+fn make_unit() {
+}
+
+fn unconstrained_return<T>() ->T {
```
Comment (bjorn3) [inline src/test/ui/never_type/never_type_lint.rs:6]:
```suggestion
fn unconstrained_return<T>() -> T {
```
nit

File: compiler/rustc_session/src/lint/builtin.rs (hunk 1 above)
Comment (nikomatsakis) [reply compiler/rustc_session/src/lint/builtin.rs:2562]: Probably, but we're still tinkering with getting it to issue warnings at the right times. That said, I think that this lint may make sense even once the transition has happened, as it represents a surprising situation most of the time.

File: src/test/ui/never_type/never_type_lint.rs (hunk 2 above)
Comment (blitzerr) [reply src/test/ui/never_type/never_type_lint.rs:6]: thanks a lot. Will make the change.

File: compiler/rustc_session/src/lint/builtin.rs [hunk 3]
```diff
@@ -2555,6 +2555,13 @@ declare_lint! {
     @feature_gate = sym::unsafe_block_in_unsafe_fn;
 }
 
+declare_lint! {
+    //FIXME: Add explanation.
+    pub FALL_BACK_TO_NEVER_TYPE,
+    Deny,
+    "Unresolved variable might fall back to never_type `!`"
```
Comment (estebank) [inline compiler/rustc_session/src/lint/builtin.rs:2562]: Nitpick: the style for user visible messages is _not_ to have leading upper-case titling.
```suggestion
    "unresolved variable might fall back to never_type `!`"
```

Comment (blitzerr) [reply compiler/rustc_session/src/lint/builtin.rs:2562]: Thanks a lot @estebank I will make the change.

**Comment by Dylan-DPC** [comment]
@blitzerr any updates on this?

**Comment by nikomatsakis** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/79366 (I've discussed this with @blitzerr)

**Pull request closed without merging**